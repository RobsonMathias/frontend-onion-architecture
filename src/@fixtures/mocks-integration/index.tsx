export * from './store.mock';
export * from './adapters.mock';
export * from './../mocks/http.mock';
export * from './../mocks/router.mock';
