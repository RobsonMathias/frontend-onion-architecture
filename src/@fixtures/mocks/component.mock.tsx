import React from 'react';

export function Mock() {
  return <div />;
}

export const mockComponent = (componentsName: string[] = []) => {
  let composed: { [key: string]: React.ReactNode } = {};
  componentsName.forEach((name: string) => {
    composed[name] = Mock;
  });
  return composed;
};
