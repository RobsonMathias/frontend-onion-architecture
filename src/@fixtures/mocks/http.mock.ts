export const mockHttp = {
  patch: jest.fn(),
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
  delete: jest.fn(),
  interceptors: {
    request: {
      use: jest.fn(),
      eject: jest.fn()
    },
    response: {
      use: jest.fn()
    }
  }
} as any;
