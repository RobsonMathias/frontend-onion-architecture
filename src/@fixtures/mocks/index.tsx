export * from './store.mock';
export * from './router.mock';
export * from './component.mock';
export * from './wrappers';
export * from './http.mock';
export * from './services.mock';
export * from './navigation.mock';
export * from './storage.mock';
export * from './use-auth.mock';
