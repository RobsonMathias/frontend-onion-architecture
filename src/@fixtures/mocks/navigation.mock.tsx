import React from 'react';

const setData = jest.fn();
const mockUseUI = jest.fn(() => ({ setData }));

const query = { get: jest.fn(), set: jest.fn() };
const set = jest.fn();
const setAll = jest.fn();
const mockUseQuery = jest.fn(() => ({ query, setAll, set }));

const replace = jest.fn();
const mockUseURL = jest.fn(() => ({ replace }));

const mockNavBar = React.forwardRef((props, ref) => <div {...props} />);

export const mockNavigation = {
  NavBar: mockNavBar,
  useQuery: mockUseQuery,
  query,
  setAll,
  set,
  useUI: mockUseUI,
  setData,
  useURL: mockUseURL,
  replace
};
