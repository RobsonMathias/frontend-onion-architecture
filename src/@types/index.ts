declare global {
  interface Window {
    localStorage: any;
    sessionStorage: any;
  }
}

export * from './task';
export * from './store';
export * from './auth';
export * from './storage';
export * from './analytics';
