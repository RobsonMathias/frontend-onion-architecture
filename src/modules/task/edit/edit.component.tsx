import React, { useEffect } from 'react';

import { TaskPayload } from '@types';
import { FieldErrors, useForm } from 'react-hook-form';
import { Grid, IconButton, TextField } from '@mui/material';
import { BiCheck, BiX } from 'react-icons/bi';

export type EditComponentProps = {
  onSubmit: (payload: TaskPayload) => void;
  onCanceled: () => void;
  loading?: boolean;
  defaultValues: TaskPayload;
};

export function EditComponent(props: EditComponentProps) {
  const { onSubmit, loading, defaultValues, onCanceled } = props;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isSubmitted }
  } = useForm({ defaultValues });

  function hasError(error?: FieldErrors): string {
    const { message } = error || { message: '' };
    return isSubmitted ? message : '';
  }

  function submit(data: TaskPayload) {
    onSubmit(data);
    reset();
  }

  function handleCanceled() {
    onCanceled();
  }

  useEffect(() => {
    return () => {
      reset();
    };
  }, []);

  return (
    <form onSubmit={handleSubmit(submit)}>
      <Grid container>
        <TextField
          variant="filled"
          type="text"
          inputProps={{ 'data-testid': 'description' }}
          label="Insira uma atividade"
          error={!!hasError(errors.description)}
          helperText={hasError(errors.description)}
          {...register('description', {
            required: 'DescriÃ§Ã£o Ã© obrigatÃ³ria.'
          })}
          sx={{ flexGrow: 1 }}
        />
        <IconButton
          aria-label="Cancelar"
          onClick={handleCanceled}
          disabled={loading}
          data-testid="cancel"
        >
          <BiX size={24} />
        </IconButton>
        <IconButton aria-label="Adicionar" disabled={loading} data-testid="submit" type="submit">
          <BiCheck size={24} />
        </IconButton>
      </Grid>
    </form>
  );
}
