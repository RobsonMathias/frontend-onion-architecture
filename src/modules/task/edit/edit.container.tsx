import React from 'react';

import { useTask } from '../hooks';
import { EditComponent } from './edit.component';
import { Task, TaskPayload } from '@types';
import { STORE_STATUS_LOADING } from 'constants/store';

export type EditContainerProps = {
  task: Task;
  onCompleted: (task?: Task) => void;
};

export function EditContainer(props: EditContainerProps) {
  const { task, onCompleted } = props;
  const { create, task: store } = useTask();

  function handleSubmit(payload: TaskPayload) {
    create(payload);
    onCompleted({ ...task, ...payload });
  }

  function handleCanceled() {
    onCompleted(undefined);
  }

  return (
    <EditComponent
      onSubmit={handleSubmit}
      onCanceled={handleCanceled}
      defaultValues={task}
      loading={store.status === STORE_STATUS_LOADING}
    />
  );
}
