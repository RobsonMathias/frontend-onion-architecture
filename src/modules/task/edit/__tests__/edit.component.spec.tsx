/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';

import { act, fireEvent, render } from '@testing-library/react';
import '@fixtures/mocks';
import React from 'react';

import { EditComponent, EditComponentProps } from '../edit.component';
import { TaskStatus } from '@types';
import { setValueForm } from './fixture';

const onCanceled = jest.fn();
const onSubmit = jest.fn();
const defaultProps: EditComponentProps = {
  defaultValues: {
    id: 'id',
    status: TaskStatus.NEW,
    description: 'description'
  },
  onSubmit,
  onCanceled
};

describe('EditComponent', () => {
  it('should render successfully', () => {
    const { container } = render(<EditComponent {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should submit form successfully', async () => {
    const { getByTestId } = render(<EditComponent {...defaultProps} />);

    await act(async () => {
      setValueForm(fireEvent, getByTestId, { description: 'new description' });
    });

    expect(onSubmit).toHaveBeenCalledWith({
      id: 'id',
      status: 'new',
      description: 'new description'
    });
  });

  it('should call onCanceled successfully', async () => {
    const { getByTestId } = render(<EditComponent {...defaultProps} />);

    await act(async () => {
      fireEvent.click(getByTestId('cancel'));
    });

    expect(onCanceled).toHaveBeenCalled();
  });

  it('should render required errors', async () => {
    const { getByTestId, container } = render(<EditComponent {...defaultProps} />);

    await act(async () => {
      setValueForm(fireEvent, getByTestId, { description: '' });
    });

    expect(container).toHaveTextContent('Descrição é obrigatória.');
  });
});
