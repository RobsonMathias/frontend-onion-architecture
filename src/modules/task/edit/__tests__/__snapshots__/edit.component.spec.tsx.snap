// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditComponent should render successfully 1`] = `
<div>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-te7hn0-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-au3a9q-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="mui-1"
          id="mui-1-label"
        >
          Insira uma atividade
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-cio0x1-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            data-testid="description"
            id="mui-1"
            name="description"
            type="text"
            value=""
          />
        </div>
        
      </div>
      <button
        aria-label="Cancelar"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="cancel"
        tabindex="0"
        type="button"
      >
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m16.192 6.344-4.243 4.242-4.242-4.242-1.414 1.414L10.535 12l-4.242 4.242 1.414 1.414 4.242-4.242 4.243 4.242 1.414-1.414L13.364 12l4.242-4.242z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Adicionar"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="submit"
        tabindex="0"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m10 15.586-3.293-3.293-1.414 1.414L10 18.414l9.707-9.707-1.414-1.414z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
